### PR TITLE
Issue #576 - use J2EEName String to find EJBDescriptor

### LINKED ISSUE #576: CDI 2.0: WebSphereEjbServicesImpl can't use InternalEjbDescriptor any more
Sometimes Weld likes to wrap things like EjbDescriptors up in it's own internal classes ... in this case one called InternalEjbDescriptor. It has a `delegate()` method on it which allowed access to the real EjbDescriptor it was wrapping. If we could pull that out then we could cast it back to our own interface and get access to additional info like the J2EEName.

In Weld 3, the InternalEjbDescriptor class has moved and been made package protected. It isn't visible to us by any normal means. We're going to have to find another way.

This is the commented out code from WebSphereEjbServicesImpl...
```
        //TODO in Weld 3, InternalEjbDescriptor is package protected and we can't use it
//        if (ejbDescriptor instanceof InternalEjbDescriptor) {
//            InternalEjbDescriptor<?> internal = (InternalEjbDescriptor<?>) ejbDescriptor;
//            descriptor = internal.delegate();
//        }
```

### PR DESCRIPTION
**Pull request by tevans78**
#build
fixes #576 
@Emily-Jiang quite a few ways to do this. The thing I don't quite like in this current version is the need to synchronize around a global map... any ideas?

### CODE REVIEW

File: dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/EjbDescriptorImpl.java [hunk 1]
```diff
@@ -225,7 +206,18 @@ public boolean equals(Object obj) {
...
     public String toString() {
-        return "EjbDescriptor: " + this.ejbJ2EENameString;
+        return PREFIX + this.ejbJ2EENameString;
     }
 
+    public static String parseJ2EEName(EjbDescriptor<?> desc) {
+        String j2eeName = null;
+        String ejbDescString = desc.toString();
+        if (ejbDescString.startsWith(PREFIX)) {
+            j2eeName = ejbDescString.substring(PREFIX.length());
```
Comment (Emily-Jiang) [inline dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/EjbDescriptorImpl.java]: I have not tested. Should this be PREFIX.length() -1 ?

File: dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/WebSphereEjbServicesImpl.java [hunk 2]
```diff
@@ -37,6 +38,15 @@
 
     private final EJBInterceptorRegistry interceptorRegistry = new EJBInterceptorRegistry();
 
+    private final Map<String, WebSphereEjbDescriptor<?>> ejbDescriptorMap;
```
Comment (Emily-Jiang) [inline dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/WebSphereEjbServicesImpl.java]: need to clear the map?

File: dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/WebSphereEjbServicesImpl.java [hunk 3]
```diff
@@ -60,29 +70,33 @@ public void registerInterceptors(EjbDescriptor<?> ejbDescriptor, InterceptorBind
...
+        String j2eeName = EjbDescriptorImpl.parseJ2EEName(descriptor);
+        WebSphereEjbDescriptor<?> webSphereEjbDescriptor = null;
+        synchronized (ejbDescriptorMap) {
+            webSphereEjbDescriptor = ejbDescriptorMap.get(j2eeName);
+        }
+        if (webSphereEjbDescriptor == null || !descriptor.equals(webSphereEjbDescriptor)) {
+            //Shouldn't be possible to get here ... the ejbDescriptorMap is global so if the descriptor isn't in there then we didn't create it
+            //the one passed in should either be exactly the same as the one in the map or it should be one wrapped by Weld.
+            //If it was wrapped by Weld then the equals method should still defer down to our original and the two should be "equal".
+            throw new IllegalStateException();
```
Comment (Emily-Jiang) [inline dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/WebSphereEjbServicesImpl.java]: need to have more message on this.

Comment (Emily-Jiang) [reply dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/WebSphereEjbServicesImpl.java]: internal error one

File: dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/EjbDescriptorImpl.java (hunk 1 above)
Comment (tevans78) [reply dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/EjbDescriptorImpl.java]: I have not tested either ... but no. If PREFIX was length 0 then you would want to start at index 0. If PREFIX was length 1 then you would want to start at index 1.

File: dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/WebSphereEjbServicesImpl.java (hunk 2 above)
Comment (tevans78) [reply dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/WebSphereEjbServicesImpl.java]: when?

Comment (Emily-Jiang) [reply dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/WebSphereEjbServicesImpl.java]: In the cleanup method

File: dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/EjbDescriptorImpl.java (hunk 1 above)
Comment (Emily-Jiang) [reply dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/EjbDescriptorImpl.java]: ok. makes sense

**Comment by Emily-Jiang** [comment]
Why did you change the file format? Is .setting checked in?

**Comment by tevans78** [comment]
The .settings are checked in. For some reason the file was previously formatted with opening brackets on the next line rather than the same line. It is "correct" now.

File: dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/WebSphereEjbServicesImpl.java (hunk 3 above)
Comment (tevans78) [reply dev/com.ibm.ws.cdi.ejb.common/src/com/ibm/ws/cdi/ejb/impl/WebSphereEjbServicesImpl.java]: Yeah I know ... I wanted a functional review before worrying about NLS.

**Comment by tevans78** [comment]
@Emily-Jiang updates made, please review ... although i still don't claim that this is elegant!
#build

**Pull request merged**